xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -315,507 +322,1272 @@ func setupWithWatch(t *testing.T) (*fakeclient.Client, *cdsBalancer, *testEDSBal
-	xdsC.InvokeWatchClusterCallback(xdsresource.ClusterUpdate{}, resourceErr)
-	// Make sure that the watch is not cancelled. This error indicates that the
-	// request cluster resource is not found. We should continue to watch it.
-	sCtx, sCancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	// Verify that the resolver error is pushed to the child policy.
+	select {
+	case err := <-resolverErrCh:
+		if !strings.Contains(err.Error(), wantClusterNACKErr) {
+			t.Fatalf("Error pushed to child policy is %v, want %v", err, wantClusterNACKErr)
+		}
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for resolver error to be pushed to the child policy")
+	}
+
+	// Remove the cluster resource from the management server, triggering a
+	// resource-not-found error.
+	resources = e2e.UpdateOptions{
+		NodeID:         nodeID,
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Verify that the watch for the cluster resource is not cancelled.
+	sCtx, sCancel = context.WithTimeout(ctx, defaultTestShortTimeout)
 	defer sCancel()
-	if _, err := xdsC.WaitForCancelClusterWatch(sCtx); err != context.DeadlineExceeded {
-		t.Fatal("cluster watch cancelled for a resource-not-found-error")
+	select {
+	case <-sCtx.Done():
+	case <-cdsResourceCanceledCh:
+		t.Fatal("Watch for cluster resource is cancelled when not expected to")
 	}
-	// Make sure the error is forwarded to the EDS balancer.
-	if err := edsB.waitForResolverError(ctx, resourceErr); err != nil {
-		t.Fatalf("Watch callback error is not forwarded to EDS balancer")
+
+	// Ensure that the ClientConn moves to TransientFailure.
+	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatalf("Timed out waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
+		}
 	}
 }
 
-// TestResolverError verifies the ResolverError() method in the CDS balancer.
+// Tests the following scenarios for resolver errors:
+//   - when a resolver error is received without any previous good update from the
+//     management server, the cds LB policy is expected to put the channel in
+//     TRANSIENT_FAILURE.
+//   - when a resolver error is received (one that is not a resource-not-found
+//     error), with a previous good update from the management server, the cds LB
+//     policy is expected to push the error down the child policy, but is expected
+//     to continue to use the previously received good configuration.
+//   - when a resolver error is received (one that is a resource-not-found error),
+//     with a previous good update from the management server, the cds LB policy
+//     is expected to push the error down the child policy and put the channel in
+//     TRANSIENT_FAILURE. It is also expected to cancel the CDS watch.
```

Comment:
Can you briefly mention that this a resolver error of type resource-not-found is due to the LDS resources being removed?

Reply:
DOne.